netty/src/main/java/io/grpc/netty/NettyServerHandler.java

```diff
@@ -787,6 +770,11 @@ public void onPingAckRead(ChannelHandlerContext ctx, long data) throws Http2Exce
           logger.log(Level.FINE, String.format("Window: %d",
               decoder().flowController().initialWindowSize(connection().connectionStream())));
         }
+      } else if (data == GRACEFUL_SHUTDOWN_PING) {
+        if (gracefulShutdown == null) {
```

Comment:
Oh, I remember why I put that comment. This is doing a check based on what is _received_. In your earlier response you said:
> It will never throw a RuntimeException unless there's a bug in the code.

But the bug could be in the _remote_ code, not in this code. That's why I had suggested a warning.